mds: Add killpoints for directory fragmentation

sidharthanup opened the pull request:
Defined Killpoints for testing completion/recovery of directory fragmentation during failure.

Fixes: https://tracker.ceph.com/issues/7320
Signed-off-by: Sidharth Anupkrishnan <sanupkri@redhat.com>
<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

batrick commented on src/common/legacy_config_opts.h:
```diff
@@ -462,6 +462,8 @@ OPTION(mds_debug_subtrees, OPT_BOOL)
 OPTION(mds_kill_export_at, OPT_INT)
 OPTION(mds_kill_import_at, OPT_INT)
+OPTION(mds_kill_dirfrag_export_at, OPT_INT)
+OPTION(mds_kill_dirfrag_import_at, OPT_INT)
```
We don't modify this file anymore.

sidharthanup replied:
Should have mentioned this before but it gave build errors when I left it out of this file.

batrick commented on qa/tasks/cephfs/test_fragment.py:
```diff
@@ -314,3 +314,88 @@ def _count_fragmented():
+    for export_killpoint in range(0, 6):
+        test_export_killpoints = make_test_killpoints(import_killpoint, export_killpoint)
+        setattr(TestKillPoints, "test_dirfrag_killpoints_%d_%d" % (import_killpoint, export_killpoint), test_export_killpoints)
+>>>>>>> f6e9cabfb0... Add testing of the dirfrag killpoints
```
rebase cruft

batrick commented on qa/tasks/cephfs/test_fragment.py:
```diff
@@ -314,3 +314,88 @@ def _count_fragmented():
+
+class TestDirfragKillPoints(TestFragmentation, CephFSTestCase):
+    # Two active MDS
+    MDSS_REQUIRED = 4
```
Why is it necessary to export/import for these tests? You should be just splitting/merging directories by modifying `mds_bal_split_size`/`mds_bal_merge_size`, right?

sidharthanup replied:
I think I got it wrong then. Since the split is between two MDS, the set of killpoints on the MDS doing the split was taken as the exporter set and the one where the frag goes is taken as the importer set. Should it just iterate over one consolidated list of killpoints with no notion of a source and dest?

sidharthanup replied:
@batrick Thoughts on this?

batrick replied:
There is a notification mechanism for splitting/merging as other MDS may have the directory fragment replicated. Testing killpoints during that notification is worthwhile.

Splitting and merging fragments do not overlap so there is no point doing an `n*m` matrix of configurations. You just need to do `n+m`.

sidharthanup replied:
Ack.

batrick commented:
retest this please

batrick commented:
@sidharthanup please rebase.

Any update on this? Ready for another round of review?

sidharthanup commented:
@batrick Just cleaned it up and pushed an updated set of killpoints which I think is more appropriate. Please do review.

sidharthanup commented:
@batrick https://pulpito.ceph.com/sidharthanup-2020-08-10_15:27:17-fs-wip-dirfrag-killpoints-distro-basic-smithi/ . Tests pass. I don't think I see anything funny happening. Please take a look.

batrick commented:
jenkins test make check

batrick commented on src/common/legacy_config_opts.h:
```diff
@@ -462,6 +462,7 @@ OPTION(mds_debug_subtrees, OPT_BOOL)
 OPTION(mds_kill_mdstable_at, OPT_INT)
 OPTION(mds_kill_export_at, OPT_INT)
 OPTION(mds_kill_import_at, OPT_INT)
+OPTION(mds_kill_dirfrag_at, OPT_INT)
```
The reason you had build errors is because you're using the old lookup mechanism:

`g_conf()->foo`

If the option is removed from this file (and eventually they all will be), then you must do:

`g_conf().get_val<int64_t>("mds_kill_dirfrag_at");`

grep around for other examples.

Please also cache the value as these lookups can be expensive in hot paths. See 6837bf5e5242631292462efe2b3599ff8ebe0d8e for an example.

sidharthanup replied:
Ahh thanks! Yea caching it is a good idea.

batrick commented on qa/tasks/cephfs/test_fragment.py:
```diff
@@ -314,3 +314,78 @@ def _count_fragmented():
+
+        self.fs.wait_for_daemons()
+
+        all_daemons = self.fs.get_daemon_names()
```
why?

sidharthanup replied:
Oh! I thought I had cleaned this up. Will remove.

batrick commented on qa/tasks/cephfs/test_fragment.py:
```diff
@@ -314,3 +314,78 @@ def _count_fragmented():
+
+    init = False
+
+    def _setup_cluster(self):
```
```suggestion
    def setUp(self):
```

sidharthanup replied:
Ack

batrick commented on qa/tasks/cephfs/test_fragment.py:
```diff
@@ -314,3 +314,78 @@ def _count_fragmented():
+def make_test_killpoints(killpointv):
+    def test_export_killpoints(self):
+        self.init = False
+        self._setup_cluster()
```
```suggestion
```

batrick commented on qa/tasks/cephfs/test_fragment.py:
```diff
@@ -314,3 +314,78 @@ def _count_fragmented():
+        if not(self._run_dir_frag(killpointv)):
+            log.error("Error for killpoint %d:%d" %(killpointv))
+        else:
+            return True
```
I'm not seeing a point in this wrapper method.

sidharthanup replied:
Yea me too. I took the export killpoints as inspiration. Will remove.

batrick commented on qa/tasks/cephfs/test_fragment.py:
```diff
@@ -314,3 +314,78 @@ def _count_fragmented():
+        rank_0_name = rank_0['name']
+        rank_1_name = rank_1['name']
+
+        command = ["config", "set", "mds_kill_dirfrag_at", str(killpointv)]
```
This is a fairly common pattern. How about you add to filesystem.py:

````
def config_set(self, opt, val, rank=0, status=None):
    ...
````

sidharthanup replied:
Ack.

batrick commented on qa/tasks/cephfs/test_fragment.py:
```diff
@@ -314,3 +314,78 @@ def _count_fragmented():
+        active_mds = self.fs.get_active_names()
+        if len(active_mds) != 2:
+            "One or more MDS did not come up"
+            return False
```
This check also looks unnecessary with the previous ones.

batrick commented on qa/tasks/cephfs/test_fragment.py:
```diff
@@ -314,3 +314,78 @@ def _count_fragmented():
+                    %(rank_0_name, rank_0_new_name, killpointv))
+
+        if status2.hadfailover_rank(self.fs.id, status, 1):
+            log.info("MDS %s crashed and active as MDS %s killpoint %d"
+                    %(rank_1_name, rank_1_new_name, killpointv))
```
```suggestion
                        self.fail(f"MDS {rank_1_name} crashed and active as MDS {rank_1_new_name} killpoint {killpointv}"
```

sidharthanup replied:
We don't want to fail the test case here right? since the mds crash and recovery was supposed to happen

batrick commented on qa/tasks/cephfs/test_fragment.py:
```diff
@@ -314,3 +314,78 @@ def _count_fragmented():
+
+        command = ["config", "set", "mds_kill_dirfrag_at", str(killpointv)]
+        result = self.fs.rank_asok(command, rank=0, status=status)
+        assert(result["success"])
```
This check is unnecessry. If the asok command fials then an exception will be raised.

sidharthanup replied:
Ack.

batrick commented on qa/tasks/cephfs/test_fragment.py:
```diff
@@ -314,3 +314,59 @@ def _count_fragmented():
             timeout=30
         )
+
+class TestDirfragKillPoints(TestFragmentation, CephFSTestCase):
```
You're double inheriting `CephFSTestCase`.

sidharthanup replied:
Ack, good catch! That was silly of me. Will push the fix.

ukernel commented:
add one kill point to MDCache::_fragment_logged

sidharthanup commented:
@ukernel Thanks! I've started testing it locally. Will push those changes.

sidharthanup commented:
Ack. I've fixed it. The setUp() method was handled incorrectly.

sidharthanup commented:
Have scheduled a run: https://pulpito.ceph.com/sidharthanup-2020-08-24_13:21:54-fs-wip-dirfrag-killpoints-distro-basic-smithi/

batrick commented:
@sidharthanup new failure, please check.

sidharthanup commented:
@batrick Yea I've refactored it to keep it under the TestFragmentation class itself rather than have a separate class for running killpoint tests. The multilevel inheritance was causing unwanted issues. Passing locally let me run teuthology again.

sidharthanup commented:
https://pulpito.ceph.com/sidharthanup-2020-08-24_20:53:38-fs-wip-dirfrag-killpoints-distro-basic-smithi/ . Tests pass.

batrick commented on qa/tasks/cephfs/test_fragment.py:
```diff
@@ -314,3 +314,41 @@ def _count_fragmented():
+
+        try:
+            self.test_oversize()
+        except Exception as e:
+            log.error(e.__str__())
```
If `test_oversize` fails, wouldn't we want that to also fail this killpoint test?

batrick commented on qa/tasks/cephfs/test_fragment.py:
```diff
@@ -314,3 +314,41 @@ def _count_fragmented():
+def make_test_killpoints(killpointv):
+    def test_export_killpoints(self):
+        self.init = False
+        CephFSTestCase.setUp(self)
```
Are you sure this is necessary?

batrick commented on qa/tasks/cephfs/test_fragment.py:
```diff
@@ -314,3 +314,41 @@ def _count_fragmented():
+        except Exception as e:
+            log.error(e.__str__())
+
+        self.fs.wait_for_daemons()
+
+        status2 = self.fs.status()
```
```suggestion
        status = self.fs.wait_for_daemons()
```

sidharthanup commented:
@batrick Thanks a lot for catching this. It's a false flag indeed! So what I could see happening from the logs and what I overlooked is that calling self.test_oversize() reinitializes the cluster as it's a unittest test method. The mds_kill_dirfrag_at config that was set is null and void now. Is rewriting the code for test_oversize() inside _run_dir_frag() the only way or do you know of a better solution to reuse test_oversize() ?

batrick commented:
> @batrick Thanks a lot for catching this. It's a false flag indeed! So what I could see happening from the teuthology logs and what I overlooked is that calling self.test_oversize() reinitializes the cluster as it's a unittest test method. The mds_kill_dirfrag_at config that was set is null and void now. Is rewriting the code for test_oversize() inside _run_dir_frag() the only way or do you know of a better solution to reuse test_oversize() ?

I don't think that's the problem, "reinitializes the cluster". But even if it was, you could convert `test_oversize` to a helper method named `_test_oversize` and have `test_oversize` call that.

batrick commented:
@sidharthanup I think you said this was ready for another review but I don't see an update?

sidharthanup commented:
@batrick Sorry. I messed up. I thought I had pushed it but not and my local clone's been deleted. I have manually done the changes again. PTAL

batrick commented:
jenkins test docs

batrick commented:
jenkins test api

batrick commented on qa/tasks/cephfs/test_fragment.py:
```diff
@@ -314,3 +317,36 @@ def _count_fragmented():
+def make_test_killpoints(killpointv):
+    def test_export_killpoints(self):
+        self.init = False
+        assert(self._run_dir_frag(killpointv))
```
```suggestion
        self._run_dir_frag(killpointv)
```

batrick commented on qa/tasks/cephfs/test_fragment.py:
```diff
@@ -314,3 +317,36 @@ def _count_fragmented():
+            log.info(f"MDS {rank_0_name} crashed and active as MDS {rank_0_new_name} killpoint {killpointv}")
+
+        self.delete_mds_coredump(rank_0_name)
+        return True
```
```suggestion
```

batrick commented on qa/tasks/cephfs/test_fragment.py:
```diff
@@ -314,3 +317,36 @@ def _count_fragmented():
+        if status2.hadfailover_rank(self.fs.id, status, 0):
+            log.info(f"MDS {rank_0_name} crashed and active as MDS {rank_0_new_name} killpoint {killpointv}")
+
+        self.delete_mds_coredump(rank_0_name)
```
move this into above `if`?

batrick reviewed (changes requested):
Please link the teuthology results too.

sidharthanup commented:
@batrick I have changed the code slightly because I noticed that the _configure() method which is called in test_oversize() : https://github.com/ceph/ceph/blob/d7e8f1211de46714758426b1e8658bb426c41a6e/qa/tasks/cephfs/test_fragment.py#L37 seems to be restarting the mds after setting the config. Why does it do the restart? My change feels a bit too hacky for me. Can you suggest a better way? (I changed it to use a status_after_configure class variable to keep track of the status)

batrick commented:
> @batrick I have changed the code slightly because I noticed that the _configure() method which is called in test_oversize() :
> 
> https://github.com/ceph/ceph/blob/d7e8f1211de46714758426b1e8658bb426c41a6e/qa/tasks/cephfs/test_fragment.py#L37-L47
> seems to be restarting the mds after setting the config. Why does it do the restart?

The old way of setting configs was to edit ceph.conf and restart the daemon. We now use either the `ceph config ...` command (via `self.config_set`) or the asok interface for transient changes (`ceph daemon mds.foo config set ....`).

>My change feels a bit too hacky for me. Can you suggest a better way? (I changed it to use a status_after_configure class variable to keep track of the status)

We can and should convert this over to use `ceph config set` via `self.config_set` but we need to fix the code in the MDS so that these configs are properly updated at runtime:

````
pdonnell@aglarond ~/ceph/src/mds$ git grep mds_bal_
CDir.cc:  int fast_limit = g_conf()->mds_bal_split_size * g_conf()->mds_bal_fragment_fast_factor;
CDir.h:    return g_conf()->mds_bal_split_size > 0 &&
CDir.h:           (int)get_frag_size() > g_conf()->mds_bal_split_size;
CDir.h:    (int)get_frag_size() < g_conf()->mds_bal_merge_size;
CInode.cc:  if (!g_conf()->mds_bal_export_pin)
MDBalancer.cc:  bal_fragment_dirs = g_conf().get_val<bool>("mds_bal_fragment_dirs");
MDBalancer.cc:  bal_fragment_interval = g_conf().get_val<int64_t>("mds_bal_fragment_interval");
MDBalancer.cc:  if (changed.count("mds_bal_fragment_dirs"))
MDBalancer.cc:    bal_fragment_dirs = g_conf().get_val<bool>("mds_bal_fragment_dirs");
MDBalancer.cc:  if (changed.count("mds_bal_fragment_interval"))
MDBalancer.cc:    bal_fragment_interval = g_conf().get_val<int64_t>("mds_bal_fragment_interval");
MDBalancer.cc:  static int num_bal_times = g_conf()->mds_bal_max;
MDBalancer.cc:  auto bal_interval = g_conf().get_val<int64_t>("mds_bal_interval");
MDBalancer.cc:  auto bal_max_until = g_conf().get_val<int64_t>("mds_bal_max_until");
MDBalancer.cc:  if (g_conf()->mds_bal_export_pin) {
MDBalancer.cc:    g_conf()->mds_bal_sample_interval) {
MDBalancer.cc:  switch(g_conf()->mds_bal_mode) {
MDBalancer.cc:    auto bal_interval = g_conf().get_val<int64_t>("mds_bal_interval");
MDBalancer.cc:    mds->mdcache->split_dir(split_dir, g_conf()->mds_bal_split_bits);
MDBalancer.cc:      if (load < target_load * (1.0 + g_conf()->mds_bal_min_rebalance)) {
MDBalancer.cc:    if (g_conf()->mds_bal_idle_threshold > 0 &&
MDBalancer.cc:  pop < g_conf()->mds_bal_idle_threshold &&
MDBalancer.cc:  if (need < amount * g_conf()->mds_bal_min_start)
MDBalancer.cc:  double needmax = need * g_conf()->mds_bal_need_max;
MDBalancer.cc:  double needmin = need * g_conf()->mds_bal_need_min;
MDBalancer.cc:  double midchunk = need * g_conf()->mds_bal_midchunk;
MDBalancer.cc:  double minchunk = need * g_conf()->mds_bal_minchunk;
MDBalancer.cc:  const bool hot = (v > g_conf()->mds_bal_split_rd && type == META_POP_IRD) ||
MDBalancer.cc:                   (v > g_conf()->mds_bal_split_wr && type == META_POP_IWR);
MDBalancer.cc:    dir_pop >= g_conf()->mds_bal_replicate_threshold) {
MDBalancer.cc:    dir_pop < g_conf()->mds_bal_unreplicate_threshold) {
MDSRank.cc:  double rate = g_conf()->mds_bal_target_decay;
MDSRank.cc:    amount = 100.0/g_conf()->mds_bal_target_decay; /* a good default for "i am trying to keep this export_target active" */
MDSRank.cc:    "mds_bal_fragment_dirs",
MDSRank.cc:    "mds_bal_fragment_interval",
Server.cc:  if (size >= g_conf()->mds_bal_fragment_size_max) {
Server.cc:    dout(10) << "fragment " << *in << " size exceeds " << g_conf()->mds_bal_fragment_size_max << " (ENOSPC)" << dendl;
````

Some are already updated to use `g_conf().get_val<foo>("config")` but some still use the old indexing method `g_conf()->config`. Those needs switched over. For hot code paths, we cache the values as you've already done for the killpoints. So you should fix those to be cached as well.

Don't forget to update the `src/common/options.cc` file to flag the options as `RUNTIME` updated.

sidharthanup commented:
@batrick is MDBalancer::handle_heartbeat called very often? i.e is it a hotpath?

sidharthanup commented on src/mds/MDSRank.cc:
```diff
@@ -645,9 +645,9 @@ void MDSRank::update_targets()
-  double rate = g_conf()->mds_bal_target_decay;
+  double rate = g_conf().get_val<double>("mds_bal_target_decay");
   if (amount < 0.0) {
-    amount = 100.0/g_conf()->mds_bal_target_decay; /* a good default for "i am trying to keep this export_target active" */
```
@batrick I noticed the variable rate wasn't reused here but mds_bal_target_decay was re-fetched in the original code. Is there is a reason for this and so should I leave this untouched?

batrick replied:
I think your fix is appropriate.

sidharthanup commented on src/mds/CInode.cc:
```diff
@@ -5398,7 +5398,7 @@ void CInode::queue_export_pin(mds_rank_t target)
 void CInode::maybe_export_pin(bool update)
 {
-  if (!g_conf()->mds_bal_export_pin)
+  if (!mdcache->mds->balancer->get_bal_export_pin())
```
@batrick I decided to use the cached value from MDBalancer here in CInode code but this change causes a build failure:
```
/home/sidharthanup/ceph/src/mds/CInode.cc: In member function ‘void CInode::maybe_export_pin(bool)’:
/home/sidharthanup/ceph/src/mds/CInode.cc:5401:30: error: invalid use of incomplete type ‘class MDBalancer’
   if (!mdcache->mds->balancer->get_bal_export_pin())
                              ^~
In file included from /home/sidharthanup/ceph/src/mds/CInode.cc:22:
/home/sidharthanup/ceph/src/mds/CDir.h:656:16: note: forward declaration of ‘class MDBalancer’
   friend class MDBalancer;
                ^~~~~~~~~~
make[2]: *** [src/mds/CMakeFiles/mds.dir/build.make:310: src/mds/CMakeFiles/mds.dir/CInode.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:5686: src/mds/CMakeFiles/mds.dir/all] Error 2
```
I didn't exactly understand the error stems from the declaration of the MDBalancer class in CDir.h?

batrick replied:
No, the MDBalancer class is incomplete because `MDBalancer.h` is not included in `CInode.cc`. You can either include it (probably not desirable for code hygiene reasons) or move the config caching to `MDSRank.h`

batrick commented:
@sidharthanup jenkins tests are failing.

sidharthanup commented:
jenkins test make check

sidharthanup commented on src/mds/MDSRank.cc:
```diff
@@ -3571,9 +3573,11 @@ MDSRankDispatcher::MDSRankDispatcher(
   : MDSRank(whoami_, fs_name_, mds_lock_, clog_, timer_, beacon_, mdsmap_,
-            msgr, monc_, mgrc, respawn_hook_, suicide_hook_, ioc)
+            msgr, monc_, mgrc, respawn_hook_, suicide_hook_, 
+	    g_conf().get_val<bool>("mds_bal_export_pin"), ioc)
```
@batrick PTAL here (and https://github.com/ceph/ceph/pull/35178/files#diff-ea4df8dd62d8f79137676b9f21d58633R490 and https://github.com/ceph/ceph/pull/35178/files#diff-89e22a97695f436e6645ce25f7bfb985R753) . It works but I'm not sure if it's proper hygiene.

batrick replied:
Why not just initialize it in the constructor body? I don't understand the reason for adding an argument to the constructor.

sidharthanup replied:
Ack. Yea that makes more sense. Updated

sidharthanup commented:
jenkins test make check

batrick commented:
@sidharthanup failing signed-off-by.

batrick commented:
Please rebase

sidharthanup commented:
Ack.

batrick commented on src/common/options.cc:
```diff
@@ -8089,6 +8089,7 @@ std::vector<Option> get_mds_options() {
 
     Option("mds_bal_export_pin", Option::TYPE_BOOL, Option::LEVEL_ADVANCED)
     .set_default(true)
+    .set_flag(Option::FLAG_RUNTIME)
```
Thanks for fixing these. If you've converted all uses of a config from `conf->foo` to `conf.get_val<bar>("foo")` then you can also remove the config from `src/common/legacy_config_opts.h`.

sidharthanup replied:
Ack. Done!

batrick commented on src/mds/MDSDaemon.cc:
```diff
@@ -747,10 +747,10 @@ void MDSDaemon::handle_mds_map(const cref_t<MMDSMap> &m)
+	timer, beacon, mdsmap, messenger, monc, &mgrc,
+	new LambdaContext([this](int r){respawn();}),
+	new LambdaContext([this](int r){suicide();}),
+	ioctx);
```
please undo this whitespace churn

sidharthanup replied:
Fixed. Thanks!

batrick commented on src/mds/Server.h:
```diff
@@ -347,6 +347,7 @@ class Server {
   double cap_revoke_eviction_timeout = 0;
   uint64_t max_snaps_per_dir = 100;
   unsigned delegate_inos_pct = 0;
+  int bal_fragment_size_max = 100000; 
```
Since this is set in the constructor, I think this should remain uninitialized.

sidharthanup replied:
Ack.

batrick reviewed (changes requested):
Also, please explain in your commit `mds: Optimize Balancer code path to u...` why the change is necessary. We discussed this here:

https://github.com/ceph/ceph/pull/35178#issuecomment-701655923

batrick commented on src/mds/MDBalancer.cc:
```diff
@@ -219,18 +243,16 @@ void MDBalancer::handle_export_pins(void)
-  static int num_bal_times = g_conf()->mds_bal_max;
-  auto bal_interval = g_conf().get_val<int64_t>("mds_bal_interval");
-  auto bal_max_until = g_conf().get_val<int64_t>("mds_bal_max_until");
+  static int num_bal_times = g_conf().get_val<int64_t>("mds_bal_max");
```
We should not have static mutable variables where possible. Please move this to `MDBalancer`.

sidharthanup replied:
Ack. I was wondering why this alone was initialized as static and so left it untouched.

batrick replied:
^ ?

batrick commented on src/mds/MDBalancer.cc:
```diff
@@ -265,7 +287,7 @@ class C_Bal_SendHeartbeat : public MDSInternalContext {
 double mds_load_t::mds_load() const
 {
-  switch(g_conf()->mds_bal_mode) {
+  switch(g_conf().get_val<int64_t>("mds_bal_mode")) {
```
This one should be cached.

batrick commented on src/mds/MDSDaemon.cc:
```diff
@@ -750,7 +750,7 @@ void MDSDaemon::handle_mds_map(const cref_t<MMDSMap> &m)
           new LambdaContext([this](int r){respawn();}),
           new LambdaContext([this](int r){suicide();}),
-	  ioctx);
+          ioctx);
```
leftover churn

batrick reviewed (changes requested):
@sidharthanup make sure all jenkins tests pass and you can then run it through teuthology for the dirfrag tests.

sidharthanup commented:
jenkins test make check

batrick commented:
status?

sidharthanup commented:
I was tinkering around with this: https://github.com/ceph/ceph/pull/35178/files#diff-729d5135082091929c032d8d6a0552bd2a5c658006aa7cb409764d85e64f9431R294 . PTAL if there's a better way to handle this.

batrick commented:
What's the problem?

sidharthanup commented:
Changed mds_load_t::mds_load() to mds_load_t::mds_load(int64_t bal_mode) since bal_mode was used inside mds_load_t::mds_load() rather than in a MDBalancer:: method. Wanted to confirm if what I did wasn't hacky.

sidharthanup commented:
jenkins test make check

sidharthanup commented:
@batrick PTAL

sidharthanup commented:
@batrick and @ukernel 
For killpoints 1-4, hitting it and subsequent mds recovery does not complete the fragmentation process. Is it poor placement of killpoints or should there be some journalling present?
For killpoints  >= 5, issue is that the fragments get merged on reconnect phase of recovery. Here's a snippet of mds log where this happens:
```
2020-11-02T21:59:18.742+0000 7f1f40fa0700  1 mds.0.17 handle_mds_map state change up:replay --> up:reconnect
2020-11-02T21:59:18.742+0000 7f1f40fa0700  1 mds.0.17 reconnect_start
2020-11-02T21:59:18.742+0000 7f1f40fa0700  1 mds.0.17 reopen_log
2020-11-02T21:59:18.742+0000 7f1f40fa0700 10 mds.0.cache rollback_uncommitted_fragments: 1 pending
2020-11-02T21:59:18.742+0000 7f1f40fa0700 10 mds.0.cache  rolling back 0x10000000000 refragment by 1 bits
2020-11-02T21:59:18.742+0000 7f1f40fa0700 10 mds.0.cache force_dir_fragment * on [inode 0x10000000000 [...2,head] /splitdir/ auth fragtree_t(*^1) v46 DIRTYPARENT f(v0 m2020-11-02T21:58:58.122628+0000 21=21+0) n(v0 rc2020-11-02T21:58:58.122628+0000 22=21+1) (iversion lock) | dirfrag=2 dirtyparent=1 dirty=1 0x562f48d05b80]
2020-11-02T21:59:18.742+0000 7f1f40fa0700 10 mds.0.cache  will combine frags under *: [0x562f48e52000,0x562f48e52480]
2020-11-02T21:59:18.742+0000 7f1f40fa0700 10 mds.0.cache adjust_dir_fragments * bits 0 srcfrags [0x562f48e52000,0x562f48e52480] on [inode 0x10000000000 [...2,head] /splitdir/ auth fragtree_t(*^1) v46 DIRTYPARENT f(v0 m2020-11-02T21:58:58.122628+0000 21=21+0) n(v0 rc2020-11-02T21:58:58.122628+0000 22=21+1) (iversion lock) | dirfrag=2 dirtyparent=1 dirty=1 0x562f48d05b80]
2020-11-02T21:59:18.742+0000 7f1f40fa0700 10 mds.0.cache  new fragtree is fragtree_t()
2020-11-02T21:59:18.742+0000 7f1f40fa0700 10 mds.0.cache.dir(0x10000000000) merge [0x562f48e52000,0x562f48e52480]
```
It starts with [rollback_uncommitted_fragments()](https://github.com/ceph/ceph/blob/master/src/mds/MDSRank.cc#L1888) being called on reconnect start which calls [force_dir_fragment()](https://github.com/ceph/ceph/blob/master/src/mds/MDCache.cc#L12268), which I believe is called due to it being an uncommitted fragment (added as one during EFragment replay), which then calls [merge]( https://github.com/ceph/ceph/blob/master/src/mds/MDCache.cc#L11229) and so we end up with one dirfrag rather than two and so the fragmentation is not completed on recovery. Thought's on how to handle this?

Also, here's a snippet of the logs during EFragment replay:
```
2020-11-02T21:59:17.754+0000 7f1f39791700 10 mds.0.journal EFragment.replay prepare 0x10000000000 * by 1
2020-11-02T21:59:17.754+0000 7f1f39791700 10 mds.0.cache add_uncommitted_fragment: base dirfrag 0x10000000000 bits 1
2020-11-02T21:59:17.754+0000 7f1f39791700 10 mds.0.cache adjust_dir_fragments * 1 on [inode 0x10000000000 [...2,head] /splitdir/ auth v44 DIRTYPARENT f(v0 m2020-11-02T21:58:58.122628+0000 21=21+0) n(v0 rc2020-11-02T21:58:58.122628+0000 22=21+1) (iversion lock) | dirfrag=1 dirtyparent=1 dirty=1 0x562f48d05b80]
2020-11-02T21:59:17.754+0000 7f1f39791700 10 mds.0.cache adjust_dir_fragments * bits 1 srcfrags [0x562f48053a80] on [inode 0x10000000000 [...2,head] /splitdir/ auth v44 DIRTYPARENT f(v0 m2020-11-02T21:58:58.122628+0000 21=21+0) n(v0 rc2020-11-02T21:58:58.122628+0000 22=21+1) (iversion lock) | dirfrag=1 dirtyparent=1 dirty=1 0x562f48d05b80]
2020-11-02T21:59:17.754+0000 7f1f39791700 10 mds.0.cache  new fragtree is fragtree_t(*^1)
2020-11-02T21:59:17.754+0000 7f1f39791700 10 mds.0.cache.dir(0x10000000000) split by 1 bits on [dir 0x10000000000 /splitdir/ [2,head] auth v=43 cv=0/0 state=1610612737|complete f(v0 m2020-11-02T21:58:58.122628+0000 21=21+0) n(v0 rc2020-11-02T21:58:58.122628+0000 21=21+0) hs=21+0,ss=0+0 dirty=21 | child=1 dirty=1 0x562f48053a80]
2020-11-02T21:59:17.754+0000 7f1f39791700 10 mds.0.cache.dir(0x10000000000)  rstatdiff n(v0 rc2020-11-02T21:58:58.122628+0000) fragstatdiff f(v0 m2020-11-02T21:58:58.122628+0000)
2020-11-02T21:59:17.754+0000 7f1f39791700 10 mds.0.cache.dir(0x10000000000)  subfrag 0* [dir 0x10000000000.0* /splitdir/ [2,head] auth v=0 cv=0/0 state=1610612737|complete f() n() hs=0+0,ss=0+0 0x562f48e52000]
2020-11-02T21:59:17.754+0000 7f1f39791700 10 mds.0.cache.dir(0x10000000000.0*) setting dir_auth=-1,-2 from -1,-2 on [dir 0x10000000000.0* /splitdir/ [2,head] auth v=0 cv=0/0 state=1610612737|complete f() n() hs=0+0,ss=0+0 0x562f48e52000]
2020-11-02T21:59:17.754+0000 7f1f39791700 15 mds.0.cache.ino(0x10000000000) maybe_export_pin update=0 [inode 0x10000000000 [...2,head] /splitdir/ auth fragtree_t(*^1) v44 DIRTYPARENT f(v0 m2020-11-02T21:58:58.122628+0000 21=21+0) n(v0 rc2020-11-02T21:58:58.122628+0000 22=21+1) (iversion lock) | dirfrag=1 dirtyparent=1 dirty=1 0x562f48d05b80]
2020-11-02T21:59:17.754+0000 7f1f39791700 10 mds.0.cache.dir(0x10000000000)  subfrag 1* [dir 0x10000000000.1* /splitdir/ [2,head] auth v=0 cv=0/0 state=1610612737|complete f() n() hs=0+0,ss=0+0 0x562f48e52480]
2020-11-02T21:59:17.754+0000 7f1f39791700 10 mds.0.cache.dir(0x10000000000.1*) setting dir_auth=-1,-2 from -1,-2 on [dir 0x10000000000.1* /splitdir/ [2,head] auth v=0 cv=0/0 state=1610612737|complete f() n() hs=0+0,ss=0+0 0x562f48e52480]
2020-11-02T21:59:17.754+0000 7f1f39791700 15 mds.0.cache.ino(0x10000000000) maybe_export_pin update=0 [inode 0x10000000000 [...2,head] /splitdir/ auth fragtree_t(*^1) v44 DIRTYPARENT f(v0 m2020-11-02T21:58:58.122628+0000 21=21+0) n(v0 rc2020-11-02T21:58:58.122628+0000 22=21+1) (iversion lock) | dirfrag=2 dirtyparent=1 dirty=1 0x562f48d05b80]
```

batrick commented:
> For killpoints 1-4, hitting it and subsequent mds recovery does not complete the fragmentation process. Is it poor placement of killpoints or should there be some journalling present?

I think the problem is your changes to `mds_bal_split_size` use `self.fs.set_config` via `self._configure`. Those configurations do not persist after the killpoint is hit because the MDS restarts. Try using `self.config_set` instead as that sets the configs through the monitors and will persist across MDS reboots.

sidharthanup commented:
> > For killpoints 1-4, hitting it and subsequent mds recovery does not complete the fragmentation process. Is it poor placement of killpoints or should there be some journalling present?
> 
> I think the problem is your changes to `mds_bal_split_size` use `self.fs.set_config` via `self._configure`. Those configurations do not persist after the killpoint is hit because the MDS restarts. Try using `self.config_set` instead as that sets the configs through the monitors and will persist across MDS reboots.

Great point! That may very well be the issue here. Let me retest.

batrick commented:
Status?

batrick commented:
Needs rebased.

sidharthanup commented:
@batrick Still facing some issues here. https://pulpito.ceph.com/sidharthanup-2020-11-17_13:10:00-fs-wip-dirfrag-killpoints15-distro-basic-smithi/ now fails with an assertion failure giving the no of fragments as 8 instead of an expected 2. I'm having some difficulty debugging this. @batrick do you see something odd in the caching changes?

batrick commented:
@sidharthanup still waiting on this change: https://github.com/ceph/ceph/pull/35178/files#r505924962

batrick commented:
rebase cruft: https://github.com/ceph/ceph/pull/35178/files#diff-277dc6e796ccecb6aa14c9357f7a86898d6ddcf4113c110a4e00e0a10f6fefa7R1233

batrick reviewed (changes requested):
I think I figured out the issue from your previous tests. The old QA suite writes out a `ceph.conf` for the MDS befor erestarting the MDS. That looks like:

````
2020-11-17T12:17:41.747 INFO:teuthology.orchestra.run.smithi105.stdout:[mds]
2020-11-17T12:17:41.747 INFO:teuthology.orchestra.run.smithi105.stdout: debug mds = 20
2020-11-17T12:17:41.747 INFO:teuthology.orchestra.run.smithi105.stdout: debug ms = 1
2020-11-17T12:17:41.747 INFO:teuthology.orchestra.run.smithi105.stdout: mds bal frag = True
2020-11-17T12:17:41.747 INFO:teuthology.orchestra.run.smithi105.stdout: mds bal fragment size max = 10000
2020-11-17T12:17:41.748 INFO:teuthology.orchestra.run.smithi105.stdout: mds bal merge size = 5
2020-11-17T12:17:41.748 INFO:teuthology.orchestra.run.smithi105.stdout: mds bal split bits = 3
2020-11-17T12:17:41.748 INFO:teuthology.orchestra.run.smithi105.stdout: mds bal split size = 100
2020-11-17T12:17:41.748 INFO:teuthology.orchestra.run.smithi105.stdout: mds debug frag = True
2020-11-17T12:17:41.748 INFO:teuthology.orchestra.run.smithi105.stdout: mds debug scatterstat = True
2020-11-17T12:17:41.748 INFO:teuthology.orchestra.run.smithi105.stdout: mds op complaint time = 180
2020-11-17T12:17:41.749 INFO:teuthology.orchestra.run.smithi105.stdout: mds verify scatter = True
2020-11-17T12:17:41.749 INFO:teuthology.orchestra.run.smithi105.stdout: osd op complaint time = 180
2020-11-17T12:17:41.749 INFO:teuthology.orchestra.run.smithi105.stdout: rados mon op timeout = 15m
2020-11-17T12:17:41.749 INFO:teuthology.orchestra.run.smithi105.stdout: rados osd op timeout = 15m
2020-11-17T12:17:41.749 INFO:teuthology.orchestra.run.smithi105.stdout: mds_bal_split_size = 100
2020-11-17T12:17:41.750 INFO:teuthology.orchestra.run.smithi105.stdout: mds_bal_merge_size = 1
````

From: /ceph/teuthology-archive/sidharthanup-2020-11-17_12:00:37-fs-wip-dirfrag-killpoints15-distro-basic-smithi/5633534/teuthology.log

This file includes settings from: 

https://github.com/ceph/ceph/blob/d7d35c2b3bd553c6d1c069d1c6d57cf669e1fbc4/qa/suites/fs/functional/overrides/frag_enable.yaml

which points to

https://github.com/ceph/ceph/blob/d7d35c2b3bd553c6d1c069d1c6d57cf669e1fbc4/qa/cephfs/overrides/frag_enable.yaml

Those `ceph.conf` settings override the `ceph config set` changes. We can actually remove

https://github.com/ceph/ceph/blob/d7d35c2b3bd553c6d1c069d1c6d57cf669e1fbc4/qa/cephfs/overrides/frag_enable.yaml#L5

and

https://github.com/ceph/ceph/blob/d7d35c2b3bd553c6d1c069d1c6d57cf669e1fbc4/qa/cephfs/overrides/frag_enable.yaml#L9

as these are the defaults (`mds_bal_frag` is actually removed.)

The other three options are not defaults so that we regularly exercise splitting/merging in the MDS during testing. These options cannot go be set through the yaml as-is anymore because they override the QA tests. I'd like to convert these configs to use `ceph config` interface but that will require broader infrastructure changes. Let's instead go back to using the legacy `self.set_conf` and restart the MDS. Please make a ticket for fixing the config setting overrides.

sidharthanup commented on src/mds/CDir.cc:
```diff
@@ -3695,7 +3695,7 @@ std::string CDir::get_path() const
 {
   // Max size a fragment can be before trigger fast splitting
-  int fast_limit = g_conf()->mds_bal_split_size * g_conf()->mds_bal_fragment_fast_factor;
+  int fast_limit = g_conf().get_val<int64_t>("mds_bal_split_size") * g_conf().get_val<double>("mds_bal_fragment_fast_factor");
```
I'm a bit confused as to where to track/cache these. It should be in MDBalancer right(As opposed to MDSRank)?

sidharthanup commented:
@batrick Thank you for figuring this out! However, we do get back to the problem of the original problem of restarting the mds to reflect the changes. As a hacky workaround, I was wondering if for now we could use both the set_config() and config_set(), to eliminate the need for a restart. See: https://github.com/ceph/ceph/pull/35178/files#diff-63705f4c68db427164408d91a6fe6e03e72db157f06904af5520283977ee237fR44-R52

batrick commented:
> @batrick Thank you for figuring this out! However, we do get back to the problem of the original problem of restarting the mds to reflect the changes. As a hacky workaround, I was wondering if for now we could use both the set_config() and config_set(), to eliminate the need for a restart. See: https://github.com/ceph/ceph/pull/35178/files#diff-63705f4c68db427164408d91a6fe6e03e72db157f06904af5520283977ee237fR44-R52

Oh right, I had forgotten why we were using `config_set` in the first place.

One easy workaround you can do is change `begin.yaml` to add:

````
...
        - libtool
        - libuuid-devel
        - xfsprogs-devel
  - ceph:
  - exec:
    mon.a:
      - ceph config set mds mds_bal_split_size foo
      - ceph config set ....
````

and then remove the `overrides/frag_enable.yaml` entirely (from all fs/multimds suites).

sidharthanup commented:
https://pulpito.ceph.com/sidharthanup-2020-11-30_13:39:08-fs-wip-dirfrag-killpoints15-distro-basic-smithi/ . Is there something wrong in how I've added the config set commands in begin.yaml?

batrick commented:
retest this please

batrick commented on qa/cephfs/begin.yaml:
```diff
@@ -50,3 +50,8 @@ tasks:
         - xfsprogs-devel
   - ceph:
+  - exec:
+    mon.a:
```
```suggestion
      mon.a:
```

batrick reviewed (changes requested):
````
2020-11-30T13:57:02.126 ERROR:teuthology.run_tasks:Saw exception from tasks.
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/run_tasks.py", line 85, in run_tasks
    ((taskname, config),) = taskdict.items()
ValueError: too many values to unpack (expected 1)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/run_tasks.py", line 87, in run_tasks
    raise RuntimeError('Invalid task definition: %s' % taskdict)
RuntimeError: Invalid task definition: {'exec': None, 'mon.a': ['ceph config set mds mds_bal_fragment_size_max 10000', 'ceph config set mds mds_bal_split_size 100', 'ceph config set mds mds_bal_merge_size 5']}
2020-11-30T13:57:02.150 ERROR:teuthology.run_tasks: Sentry event: https://sentry.ceph.com/organizations/ceph/?query=ac49d8d2981c44dcad905f255b1246c2
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/run_tasks.py", line 85, in run_tasks
    ((taskname, config),) = taskdict.items()
ValueError: too many values to unpack (expected 1)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/run_tasks.py", line 87, in run_tasks
    raise RuntimeError('Invalid task definition: %s' % taskdict)
RuntimeError: Invalid task definition: {'exec': None, 'mon.a': ['ceph config set mds mds_bal_fragment_size_max 10000', 'ceph config set mds mds_bal_split_size 100', 'ceph config set mds mds_bal_merge_size 5']}
````

sidharthanup commented:
retest this please

sidharthanup commented:
jenkins test make check

batrick commented:
````
flake8 run-test: commands[0] | flake8 --select=F,E9 --exclude=venv,.tox
./tasks/cephfs/test_fragment.py:67:28: F821 undefined name 'mds_kill_dirfrag_at'
ERROR: InvocationError for command /home/jenkins-build/build/workspace/ceph-pull-requests/qa/.tox/flake8/bin/flake8 --select=F,E9 --exclude=venv,.tox (exited with code 1)
````

batrick commented:
@sidharthanup please run through QA again when finished.

sidharthanup commented:
Ack I forgot to push some of the latest fixes.

sidharthanup commented:
jenkins test make check

sidharthanup commented:
@batrick Pls review again. https://pulpito.ceph.com/sidharthanup-2020-12-17_14:01:08-fs-wip-dirfrag-killpoints17-distro-basic-smithi/ seems to be hitting the same assertion failure as before.

batrick commented:
> @batrick Pls review again. https://pulpito.ceph.com/sidharthanup-2020-12-17_14:01:08-fs-wip-dirfrag-killpoints17-distro-basic-smithi/ seems to be hitting the same assertion failure as before.

It looks like the `begin.yaml` changes were lost?

https://github.com/ceph/ceph/pull/35178#issuecomment-732306662

Please let me know what you fixed/changed.

sidharthanup commented:
Uh oh.. Not sure why the begin.yaml and the frag_enable deletes got lost. Have reupdated. Waiting for teuthology.

batrick commented:
jenkins test make check

batrick commented:
Please rebase

sidharthanup commented on src/common/options/mds.yaml.in:
```diff
@@ -961,6 +961,12 @@ options:
   fmt_desc: Ceph will inject MDS failure in the subtree import code
     (for developers only).
   with_legacy: true
+- name: mds_kill_dirfrag_at
```
@batrick with the change in the config format introduced a while back, have I updated this correctly?

batrick replied:
add runtime flag too,  otherwise looks okay to me.

batrick commented:
Now: https://github.com/ceph/ceph/pull/48130
